gnucash/gnome-utils/gnc-account-sel.c

```diff
@@ -136,151 +227,453 @@ gnc_account_sel_class_init (GNCAccountSelClass *klass)
+                            const char* c;
+                            gunichar uc;
+                            gboolean found = FALSE;
+
+                            c = g_utf8_offset_to_pointer (account_full_name, item_offset_to_sep_char);
+                            item_offset_to_sep_char++;
+
+                            while (*c)
+                            {
+                                uc = g_utf8_get_char (c);
+                                if (uc == sep_char)
+                                {
+                                    found = TRUE;
+                                    break;
+                                }
+                                c = g_utf8_next_char (c);
+                                item_offset_to_sep_char ++;
+                            }
+
+                            if (found && (item_offset_to_sep_char < prefix_len))
+                            {
+                                prefix_len = item_offset_to_sep_char;
+                                memset (gas->prefix, 0, BUFLEN);
+                                g_utf8_strncpy (gas->prefix, account_full_name, prefix_len);
+                            }
+
+                            if (found && (item_offset_to_sep_char == prefix_len))
+                            {
+                                char tmp_prefix[BUFLEN];
+
+                                memset (tmp_prefix, 0, BUFLEN);
+                                g_utf8_strncpy (tmp_prefix, account_full_name, prefix_len);
+
+                                if (g_strcmp0 (gas->prefix, tmp_prefix) != 0)
+                                {
+                                    do
+                                    {
+                                        gchar *tmp = g_strdup (gas->prefix);
+                                        prefix_len--;
+
+                                        memset (tmp_prefix, 0, BUFLEN);
+                                        g_utf8_strncpy (tmp_prefix, account_full_name, prefix_len);
+                                        memset (gas->prefix, 0, BUFLEN);
+                                        g_utf8_strncpy (gas->prefix, tmp, prefix_len);
+                                        g_free (tmp);
+
+                                    } while (g_strcmp0 (gas->prefix, tmp_prefix) != 0);
+                                }
+                            }
+                        }
+                        g_free (lower_account_full_name);
+                    }
+                    g_free (normal_account_full_name);
+                }
+                g_free (account_full_name);
+                valid = gtk_tree_model_iter_next (fmodel, &iter);
+            }
+            if (gas->prefix[0] == 0)
+                g_utf8_strncpy (gas->prefix, lower_entered_text, entered_len);
 
-    gas->initDone = TRUE;
+            g_free (lower_entered_text);
+        }
+        g_free (normal_entered_text);
+    }
```

Comment:
Everything between the previous comment and this one needs a refactor to be more intelligible.  The following is my attempt, please check to make sure that it's what you intended:
```
static char*
normalize_and_lower (char* utf8_string)
{
    char *normalized, *lowered;
    g_return_val_if_fail (utf8_string && *utf8_string, NULL);

    normalized = g_utf8_normalize (utf8_string, -1, G_NORMALIZE_ALL);
    g_free (utf8_string);
    if (!normalized)
        return NULL;
    lowered = g_utf8_strdown (normalized, -1);
    g_free (normalized);
    return lowered;
}

static inline gboolean
find_next_separator (const char* c, gunichar sep_char,
    size_t &offset)
{
    while (*c)
    {
        gboolean found = FALSE;
        gunichar uc = g_utf8_get_char (c);
        if (uc == sep_char)
        {
            found = TRUE;
            break;
        }
        c = g_utf8_next_char (c);
        ++(*offset);
    }
    return found;
}

/* Set gas->prefix from the account full name either to the next separator or
 * to the 
 * account_full_name is normalized and lowercased.
 */
static void
set_prefix_from_account_name (char* account_full_name, GNCAccountSel *gas,
                              int entered_len, gunichar sep_char, gint *prefix_len)
{
    gint item_offset_to_sep_char = entered_len;
    const char* c;
    gboolean found = FALSE;

    c = g_utf8_offset_to_pointer (account_full_name, item_offset_to_sep_char);
    item_offset_to_sep_char++;
    found = find_next_separator (c, sep_char, &item_offset_to_sep_char);

    if (!found)
        return;

    if (item_offset_to_sep_char < prefix_len)
    {
        *prefix_len = item_offset_to_sep_char;
        memset (gas->prefix, 0, BUFLEN);
        g_utf8_strncpy (gas->prefix, account_full_name, *prefix_len);
    }

    if (item_offset_to_sep_char == prefix_len)
    {
        char tmp_prefix[BUFLEN];

        memset (tmp_prefix, 0, BUFLEN);
        g_utf8_strncpy (tmp_prefix, account_full_name, prefix_len);

        while (g_strcmp0 (gas->prefix, tmp_prefix) != 0)
        {
            gchar *tmp = g_strdup (gas->prefix);
            prefix_len--;

            memset (tmp_prefix, 0, BUFLEN);
            g_utf8_strncpy (tmp_prefix, account_full_name, prefix_len);
            memset (gas->prefix, 0, BUFLEN);
            g_utf8_strncpy (gas->prefix, tmp, prefix_len);
            g_free (tmp);
        }
    }
}

static void
entry_insert_text_cb (GtkEntry *entry, const gchar *text, gint length,
                      gint *position, gpointer user_data)
 {
     GNCAccountSel *gas = GNC_ACCOUNT_SEL(user_data);
     GtkTreeModel *fmodel = gtk_combo_box_get_model (GTK_COMBO_BOX(gas->combo));
     const gchar *sep_char = gnc_get_account_separator_string ();
     gchar *entered_text;
     glong entered_len;
     gint prefix_len = G_MAXINT;
     gunichar sep_uchar = g_utf8_get_char (normalize_and_lower (sep_char));
     GtkTreeIter iter;
     gboolean valid;

     if (g_strcmp0 (text, sep_char) != 0)
         return;

     memset (gas->prefix, 0, BUFLEN);

     entered_text =
         normalize_and_lower (gtk_editable_get_chars (GTK_EDITABLE(entry), 0, -1));

     if (!entered_text)
         return;

     entered_len = g_utf8_strlen (entered_text, -1); //characters
     // Get the first item in the list
     valid = gtk_tree_model_get_iter_first (fmodel, &iter);

     // Walk through the list, reading each full name
     while (valid)
     {
         gchar *account_full_name, *lower_account_full_name;

         gtk_tree_model_get (fmodel, &iter, ACCT_COL_NAME, &account_full_name, -1);
         account_full_name = normalize_and_lower (account_full_name);

         if (!account_full_name)
             continue;

         if (g_str_has_prefix (account_full_name, entered_text))
             set_prefix_from_account_name (account_full_name, gas,
                                           entered_len, sep_char);

         g_free (account_full_name);
         valid = gtk_tree_model_iter_next (fmodel, &iter);
     }

     if (gas->prefix[0] == 0)
         g_utf8_strncpy (gas->prefix, entered_text, entered_len);

     g_free (entered_text);

     if (gas->prefix[0] != 0)
     {
         g_signal_handlers_block_by_func (GTK_EDITABLE(entry), (gpointer) entry_insert_text_cb, user_data);
         gtk_editable_delete_text (GTK_EDITABLE(entry), 0, -1);
         gtk_editable_set_position (GTK_EDITABLE(entry), 0);
         gtk_editable_insert_text (GTK_EDITABLE(entry), gas->prefix, -1, position);
         g_signal_handlers_unblock_by_func (GTK_EDITABLE(entry), (gpointer) entry_insert_text_cb, user_data);
         g_signal_stop_emission_by_name (GTK_EDITABLE(entry), "insert_text");
     }
 }
```

Reply:
I see what you mean, I have a tendency to just add parts when I come across I need them. Will try and do more refactoring when I push.